File: json/json-annotation-processor/src/main/java/ru/tinkoff/kora/json/annotation/processor/JsonUtils.java
package ru.tinkoff.kora.json.annotation.processor;

import ru.tinkoff.kora.annotation.processor.common.AnnotationUtils;
import ru.tinkoff.kora.annotation.processor.common.CommonUtils;
import ru.tinkoff.kora.annotation.processor.common.ProcessingErrorException;

import javax.annotation.Nullable;
import javax.lang.model.element.Element;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.util.List;


public class JsonUtils {
    public static String jsonClassPackage(Elements elements, Element typeElement) {
        return elements.getPackageOf(typeElement).getQualifiedName().toString();
    }

    public static String jsonWriterName(Element typeElement) {
        return CommonUtils.getOuterClassesAsPrefix(typeElement) + typeElement.getSimpleName() + "JsonWriter";
    }

    public static String jsonWriterName(Types types, TypeMirror typeMirror) {
        var typeElement = types.asElement(typeMirror);

        return jsonWriterName(typeElement);
    }

    public static String jsonReaderName(TypeElement typeElement) {
        return CommonUtils.getOuterClassesAsPrefix(typeElement) + typeElement.getSimpleName() + "JsonReader";
    }

    public static String jsonReaderName(Types types, TypeMirror typeMirror) {
        var typeElement = types.asElement(typeMirror);

        return jsonReaderName((TypeElement) typeElement);
    }


    @Nullable
    public static String discriminatorField(Types types, TypeElement element) {
        if (element.getModifiers().contains(Modifier.SEALED)) {
            var annotation = AnnotationUtils.findAnnotation(element, JsonTypes.jsonDiscriminatorField);
            if (annotation != null) {
                return AnnotationUtils.parseAnnotationValueWithoutDefault(annotation, "value");
            }
        }
        var superClass = types.asElement(element.getSuperclass());
        if (superClass instanceof TypeElement && superClass.getModifiers().contains(Modifier.SEALED)) {
            var annotation = AnnotationUtils.findAnnotation(superClass, JsonTypes.jsonDiscriminatorField);
            if (annotation != null) {
                return AnnotationUtils.parseAnnotationValueWithoutDefault(annotation, "value");
            }
        }
        for (var directSupertype : types.directSupertypes(element.asType())) {
            if (directSupertype.toString().equals("java.lang.Object")) {
                continue;
            }
            var superelement = types.asElement(directSupertype);
            var discriminator = discriminatorField(types, (TypeElement) superelement);
            if (discriminator != null) {
                return discriminator;
            }
        }
        return null;
    }

    public static List<String> discriminatorValue(TypeElement element) {
        var annotation = AnnotationUtils.findAnnotation(element, JsonTypes.jsonDiscriminatorValue);
        if (annotation != null) {
            var value = AnnotationUtils.<List<String>>parseAnnotationValueWithoutDefault(annotation, "value");
            if (value != null) {
                if (value.isEmpty()) {
                    throw new ProcessingErrorException("Discriminator value can't be empty array", element, annotation);
                }
                return value;
            }
        }
        return List.of(element.getSimpleName().toString());
    }
}

File: json/json-symbol-processor/src/main/kotlin/ru/tinkoff/kora/json/ksp/JsonUtils.kt
package ru.tinkoff.kora.json.ksp

import com.google.devtools.ksp.getAllSuperTypes
import com.google.devtools.ksp.symbol.*
import com.squareup.kotlinpoet.STAR
import com.squareup.kotlinpoet.TypeName
import com.squareup.kotlinpoet.ksp.toTypeVariableName
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotation
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findValue
import ru.tinkoff.kora.ksp.common.generatedClassName
import ru.tinkoff.kora.ksp.common.isJavaRecord
import java.util.*

fun isSealed(classDeclaration: KSClassDeclaration): Boolean {
    return classDeclaration.modifiers.contains(Modifier.SEALED)
}

fun jsonClassPackage(classDeclaration: KSClassDeclaration): String {
    return classDeclaration.packageName.asString()
}

fun KSClassDeclaration.jsonReaderName() = this.generatedClassName("JsonReader")
fun KSClassDeclaration.jsonWriterName() = this.generatedClassName("JsonWriter")

fun findJsonField(param: KSValueParameter, jsonClass: KSClassDeclaration): KSAnnotation? {
    if (jsonClass.isJavaRecord()) {
        return jsonClass.getAllFunctions()
            .firstOrNull { it.simpleName.asString() == param.name!!.asString() && it.parameters.isEmpty() }
            ?.findAnnotation(JsonTypes.jsonFieldAnnotation)
    }
    return findJsonField(param)
}

fun findJsonField(param: KSAnnotated): KSAnnotation? {
    return param.findAnnotation(JsonTypes.jsonFieldAnnotation)
}


fun KSClassDeclaration.discriminatorField(): String? {
    if (this.packageName.asString() == "kotlin") {
        return null
    }
    if (this.modifiers.contains(Modifier.SEALED)) {
        val annotation = this.findAnnotation(JsonTypes.jsonDiscriminatorField)
        if (annotation != null) {
            return annotation.findValue<String>("value")
        }
    }
    for (type in this.superTypes) {
        val supertype = type.resolve().declaration as KSClassDeclaration
        val discriminator = supertype.discriminatorField()
        if (discriminator != null) {
            return discriminator
        }
    }
    return null
}

fun KSClassDeclaration.discriminatorValues(): List<String> {
    return findAnnotation(JsonTypes.jsonDiscriminatorValue)
        ?.findValue<List<String>>("value")
        ?: listOf(this.simpleName.asString())
}


fun detectSealedHierarchyTypeVariables(jsonClassDeclaration: KSClassDeclaration, subclasses: List<KSClassDeclaration>): Pair<IdentityHashMap<KSTypeParameter, TypeName>, ArrayList<TypeName>> {
    val subclassesSupertypes = subclasses.map { it.getAllSuperTypes().filter { it.declaration == jsonClassDeclaration }.first() }
    val typeNames = ArrayList<TypeName>(jsonClassDeclaration.typeParameters.size)
    val map = IdentityHashMap<KSTypeParameter, TypeName>()

    for ((index, ksTypeParameter) in jsonClassDeclaration.typeParameters.withIndex()) {
        var typeName = null as TypeName?
        val subtypeArgs = ArrayList<KSTypeParameter>()
        for ((subclassIndex, subclass) in subclasses.withIndex()) {
            val supertype = subclassesSupertypes[subclassIndex]
            val supertypeArgument = supertype.arguments[index].type!!.resolve()
            val argDeclaration = supertypeArgument.declaration
            if (argDeclaration !is KSTypeParameter) {
                typeName = STAR
                continue
            }
            val subtypeArgument = subclass.typeParameters.filter { it.name == argDeclaration.name }.firstOrNull()
            if (subtypeArgument == null) {
                typeName = STAR
                continue
            }
            if (typeName == null || typeName != STAR) {
                subtypeArgs.add(subtypeArgument)
                typeName = if (subtypeArgument.bounds.toList() != ksTypeParameter.bounds.toList()) {
                    STAR
                } else {
                    ksTypeParameter.toTypeVariableName()
                }
            }
        }
        typeNames.add(typeName!!)
        subtypeArgs.forEach { map[it] = typeName }
    }
    return map to typeNames
}

